impress-org/givewp
# Feature: donation confirmation page redirect

Resolves [GIVE-704]

## Description



This adds the ability for a donation form to redirect to the success page defined in GiveWP global settings. 

In addition to the form builder settings, more flexibility can be applied using the filters below:

Global

```php
add_filter( 'givewp_donation_confirmation_page_redirect_enabled', '__return_true' );
```

Per-form enable

```php
add_filter('givewp_donation_confirmation_page_redirect_enabled', function ($shouldRedirect, $formId) {
    if ($formId === 1) {
        $shouldRedirect = true;
    }
    
    return $shouldRedirect;
}, 10, 2);
```

Per-form permalink
```php
add_filter('givewp_donation_confirmation_page_redirect_permalink', function ($page, $formId) {
    if ($pageId = give()->form_meta->get_meta($formId, 'donation_confirmation_page_redirect_id', true)) {
        $page = get_permalink($pageId);
    }

    return $page;
}, 10, 2);
```

## Affects



This affects the donation confirmation page rendering of the give_receipt shortcode.  Since this is a shared page and shortcode with v2 and v3 forms, we'll want to make sure v2 forms are still working as expected along the way.

The approach taken was to add carefully placed actions and filters in the existing v2 logic so we can hook into them for the purposes of v3 forms. These include:

- `give_donation_confirmation_page_enqueue_scripts`
- `give_receipt_shortcode_output`
- `give_donation_confirmation_success_page_shortcode_view`

## Visuals



Demo Video:
https://www.loom.com/share/eed72b9765a44301afa5956178badfbd?sid=535e20ea-4417-4609-8fe9-e92f41b36aa1

<img width="1512" alt="Screenshot 2024-08-02 at 1 57 13 PM" src="https://github.com/user-attachments/assets/97ec3d20-a32f-482a-9a94-31fd7fc27110">

<img width="1504" alt="Screenshot 2024-07-10 at 10 38 24 AM" src="https://github.com/impress-org/givewp/assets/10138447/bb0342fd-9bff-4a31-9438-d5eca7997b4f">

## Testing Instructions



ZIP: https://github.com/impress-org/givewp/actions/runs/10219452643

**Form settings**
- Create a v3 form
- Open form settings and enable confirmation page redirect setting
- Make sure you have a confirmation page defined in give settings
- Donate to the form and make sure it redirects to that confirmation page 🎉 

**Test the filters**

- Add filter `add_filter( 'givewp_donation_confirmation_page_redirect', '__return_true' );`
- Make sure a success page exists in GiveWP settings with `give_receipt` shortcode
- Donate using a v3 form

## Pre-review Checklist






[GIVE-704]: https://stellarwp.atlassian.net/browse/GIVE-704?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
